File: android/app/src/main/kotlin/com/osasuwu/like_spotify/LikeDestination.kt
package com.osasuwu.like_spotify

/**
 * Where a like goes: the service's own "liked songs", a playlist of the
 * user's own, or both.
 *
 * The Kotlin twin of the Dart `LikeDestination`; [id] is the same string on
 * both sides, so the value Dart writes to SharedPreferences reads back here
 * unchanged. Deliberately pure — no Android API — so both native like paths
 * (the Spotify worker and the YouTube Music liker) share one implementation
 * of the branching, and a JUnit test can reach it.
 */
enum class LikeDestination(val id: String) {
    NATIVE("native"),
    PLAYLIST("playlist"),
    BOTH("both");

    /** Whether the service's own like should be sent. */
    val likesNatively: Boolean get() = this != PLAYLIST

    /** Whether the track should be added to the user's like playlist. */
    val addsToPlaylist: Boolean get() = this != NATIVE

    companion object {
        /** New installs, and installs that predate the setting, like natively. */
        val DEFAULT = NATIVE

        /** Unknown or missing ids fall back to [DEFAULT]. */
        fun fromId(id: String?): LikeDestination =
            values().firstOrNull { it.id == id } ?: DEFAULT

        /**
         * The destination to actually run, given the configured playlist name.
         *
         * A playlist destination with no playlist name has nowhere to put the
         * song. The settings screen refuses to save such a config, so this only
         * catches one written by an older build or by hand: rather than failing
         * every like, it falls back to the service's own like.
         */
        fun resolve(id: String?, playlistName: String): LikeDestination {
            val destination = fromId(id)
            return if (destination.addsToPlaylist && playlistName.isBlank()) NATIVE else destination
        }

        /**
         * Whether the like as a whole counted.
         *
         * A one-leg destination is exactly as good as its leg. [BOTH] wants the
         * song liked in two places but one is enough for it to be liked at all,
         * so it only fails when both legs fail.
         */
        fun succeeded(destination: LikeDestination, nativeOk: Boolean, playlistOk: Boolean): Boolean =
            when (destination) {
                NATIVE -> nativeOk
                PLAYLIST -> playlistOk
                BOTH -> nativeOk || playlistOk
            }
    }
}

File: lib/domain/entities/like_destination.dart
/// Where a like goes: the service's own "liked songs", a playlist of your
/// own, or both.
///
/// YouTube Music is the reason this exists — liked songs and liked videos
/// share one bucket there, so a liked song lands among everything else the
/// account has ever liked. Sending likes to a playlist instead gives them a
/// place of their own. Spotify has the same setting so the two services
/// behave the same way.
///
/// Deliberately one global setting rather than one per service, the same way
/// the archive and best playlist names are global. The Kotlin side keeps a
/// matching enum in `LikeDestination.kt`.
enum LikeDestination {
  native(id: 'native', displayName: 'Liked songs'),
  playlist(id: 'playlist', displayName: 'A playlist'),
  both(id: 'both', displayName: 'Both');

  const LikeDestination({required this.id, required this.displayName});

  /// Stable identifier used for persistence and the platform channel.
  final String id;

  /// Human-readable name for the UI.
  final String displayName;

  /// New installs and existing installs that predate the setting both keep
  /// the behaviour the app has always had: the service's own like.
  static const LikeDestination defaultDestination = LikeDestination.native;

  /// Whether the service's own like should be sent.
  bool get likesNatively => this != LikeDestination.playlist;

  /// Whether the track should be added to the user's like playlist.
  bool get addsToPlaylist => this != LikeDestination.native;

  /// Resolves a persisted [id]; unknown or missing values fall back to
  /// [defaultDestination] so an older build never crashes on a newer value.
  static LikeDestination fromId(String? id) {
    for (final destination in LikeDestination.values) {
      if (destination.id == id) return destination;
    }
    return defaultDestination;
  }

  /// The destination to actually run, given the configured playlist name.
  ///
  /// A playlist destination with no playlist name has nowhere to put the
  /// song. `validate()` stops the user saving such a config, so this only
  /// catches a hand-edited or partially written one: rather than failing
  /// every like, it falls back to the service's own like.
  static LikeDestination resolve(LikeDestination destination, String playlistName) {
    if (destination.addsToPlaylist && playlistName.trim().isEmpty) {
      return LikeDestination.native;
    }
    return destination;
  }
}
